crates/polars-ops/src/series/ops/cum_agg.rs

```diff
@@ -0,0 +1,272 @@
+    };
+    out.with_name(ca.name())
+}
+
+fn cummin_numeric<T>(ca: &ChunkedArray<T>, reverse: bool) -> ChunkedArray<T>
+where
+    T: PolarsNumericType,
+    ChunkedArray<T>: FromIterator<Option<T::Native>>,
+{
+    let init = Bounded::max_value();
+    let out: ChunkedArray<T> = match reverse {
+        false => ca.into_iter().scan(init, det_min).collect_trusted(),
+        true => ca.into_iter().rev().scan(init, det_min).collect_reversed(),
+    };
+    out.with_name(ca.name())
+}
+
+fn cumsum_numeric<T>(ca: &ChunkedArray<T>, reverse: bool) -> ChunkedArray<T>
+where
+    T: PolarsNumericType,
+    ChunkedArray<T>: FromIterator<Option<T::Native>>,
+{
+    let init = None;
+    let out: ChunkedArray<T> = match reverse {
+        false => ca.into_iter().scan(init, det_sum).collect_trusted(),
+        true => ca.into_iter().rev().scan(init, det_sum).collect_reversed(),
+    };
+    out.with_name(ca.name())
+}
+
+fn cumprod_numeric<T>(ca: &ChunkedArray<T>, reverse: bool) -> ChunkedArray<T>
+where
+    T: PolarsNumericType,
+    ChunkedArray<T>: FromIterator<Option<T::Native>>,
+{
+    let init = None;
+    let out: ChunkedArray<T> = match reverse {
+        false => ca.into_iter().scan(init, det_prod).collect_trusted(),
+        true => ca.into_iter().rev().scan(init, det_prod).collect_reversed(),
+    };
+    out.with_name(ca.name())
+}
+
+/// Get an array with the cumulative product computed at every element.
+///
+/// If the [`DataType`] is one of `{Int8, UInt8, Int16, UInt16, Int32, UInt32}` the `Series` is
+/// first cast to `Int64` to prevent overflow issues.
+pub fn cumprod(s: &Series, reverse: bool) -> PolarsResult<Series> {
+    use DataType::*;
+    let out = match s.dtype() {
+        Boolean | Int8 | UInt8 | Int16 | UInt16 | Int32 | UInt32 => {
+            let s = s.cast(&Int64)?;
+            cumprod_numeric(s.i64()?, reverse).into_series()
+        },
+        Int64 => cumprod_numeric(s.i64()?, reverse).into_series(),
+        UInt64 => cumprod_numeric(s.u64()?, reverse).into_series(),
+        Float32 => cumprod_numeric(s.f32()?, reverse).into_series(),
+        Float64 => cumprod_numeric(s.f64()?, reverse).into_series(),
+        dt => polars_bail!(opq = cumprod, dt),
+    };
+    Ok(out)
+}
+
+/// Get an array with the cumulative sum computed at every element
+///
+/// If the [`DataType`] is one of `{Int8, UInt8, Int16, UInt16}` the `Series` is
+/// first cast to `Int64` to prevent overflow issues.
+pub fn cumsum(s: &Series, reverse: bool) -> PolarsResult<Series> {
+    use DataType::*;
+    let out = match s.dtype() {
+        Boolean => {
+            let s = s.cast(&UInt32)?;
+            cumsum_numeric(s.u32()?, reverse).into_series()
+        },
+        Int8 | UInt8 | Int16 | UInt16 => {
+            let s = s.cast(&Int64)?;
+            cumsum_numeric(s.i64()?, reverse).into_series()
+        },
+        dt if dt.is_numeric() => {
+            with_match_physical_numeric_polars_type!(s.dtype(), |$T| {
```

Comment:
We shoul make a macro for the remaining numeric dtypes. Otherwise this code will implement for the int8,uint8,int16,uint16, where it is wasteful to compile those branches as we will never hit them.